pallets/dkg-metadata/src/lib.rs

```diff
@@ -1204,6 +1234,90 @@ pub mod pallet {
 			Ok(().into())
 		}
 	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		/// Validate unsigned call to this module.
+		///
+		/// By default unsigned transactions are disallowed, but implementing the validator
+		/// here we make sure that some particular calls (the ones produced by offchain worker)
+		/// are being whitelisted and marked as valid.
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			// Now let's check if the transaction has any chance to succeed.
+			let current_block = <frame_system::Pallet<T>>::block_number();
+			let next_unsigned_at = <NextUnsignedAt<T>>::get();
+			if next_unsigned_at > current_block {
+				frame_support::log::debug!(
+					target: "runtime::dkg_metadata",
+					"validate unsigned: early block: current: {:?}, next_unsigned_at: {:?}",
+					current_block,
+					next_unsigned_at,
+				);
+				return InvalidTransaction::Stale.into()
+			}
+			// Next, let's check that we call the right function.
+			// Here we will use match stmt, to match over the call and see if it is
+			// one of the functions we allow. if not we should return
+			// `InvalidTransaction::Call.into()`.
+			// we should handle the following calls:
+			// 1. `submit_public_key`
+			// 2. `submit_next_public_key`
+			// 3. `submit_public_key_signature`
+			// 4. `submit_misbehaviour_reports`.
+			// other than that we should return `InvalidTransaction::Call.into()`.
+			let is_valid_call = matches! {
+				call,
+				Call::submit_public_key { .. } |
+					Call::submit_next_public_key { .. } |
+					Call::submit_public_key_signature { .. } |
+					Call::submit_misbehaviour_reports { .. }
```

Comment:
How about adding a check here to ensure the sender is a member of current authorities?

Reply:
We need checks like that everywhere to ensure this doesn't lead to spam attacks.

Reply:
The spam attacks are prevented already by the check above that, but yes I agree, we will have to move the checks from the xt to a helper function and call it here and also in the xt itself again.

Reply:
> How about adding a check here to ensure the sender is a member of current authorities?

Reply:
I did add a check to only allow these unsigned tx to be from the offchain workers, by checking the source, if it is not local, it will be rejected before it reaches the txpool.